src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -301,6 +302,12 @@ create_reference_table(PG_FUNCTION_ARGS)
 	 */
 	EnsureRelationKindSupported(relationId);
 
+	/*
+	 * Take an exclusive lock on pg_dist_node to serialize pg_dist_node changes
+	 * with reference table distribution
+	 */
+	LockRelationOid(DistNodeRelationId(), AccessShareLock);
```

Comment:
AccessShareLock does not seem like it would resolve the issue, since it only conflicts with AccessExclusiveLock.
https://www.postgresql.org/docs/current/explicit-locking.html

In master_add_node we take an ExclusiveLock. Here we want a lock that conflicts with ExclusiveLock, but not with itself.